### DISCUSSION TITLE
Piccolo admin without auth

### DISCUSSION THREAD

**Discussion by chilico** [piccolo-orm/piccolo_admin#412] [Q&A]
We have multiple distributed cloud services, some with their own db with a Piccolo layer. Everything is locked down with our own auth, and we would like the Piccolo admin(s) to be completely open so that we can seamlessly route between them. Ideally we would do without the BaseUser table altogether - we simply don't need it. Is this somehow possible?

**Comment by sinisaos** [comment]
@chilico I don't think you can do that now, but Piccolo Admin is protected by `SessionsAuthBackend`, which has the `allow_unauthenticated` argument. In your local Piccolo Admin installation you need to change this

https://github.com/piccolo-orm/piccolo_admin/blob/2308515b3cbaa3cae72f7dda70e8b9818760011c/piccolo_admin/endpoints.py#L790-L795

 to this

```python
backend=SessionsAuthBackend(
    allow_unauthenticated=True,
),
```

After these changes you can access Piccolo Admin without providing a user. This can be configured via `create_admin`, but we should wait for @dantownsend to say what he thinks about it.

**Reply by dantownsend** [reply]
@sinisaos That's an interesting idea - I hadn't considered using `allow_unauthenticated=True`.

I tried it locally, and it works surprisingly well.

I think the reason we stalled on this in the past was because bits of the Piccolo Admin UI assume a user is logged in (e.g. we show their username, and we have a logout button in the nav bar dropdown). For now though, we could ignore those problems.

In terms of exposing the `allow_unauthenticated=True` option in `create_admin`, there are a few options:

1. Just add `allow_authenticated` as an argument - but there are already loads of arguments.
2. What I would ideally like to do is remove a bunch of the session related ones, and aggregate them somehow (maybe just let a user pass in a custom `SessionsAuthBackend` directly). The problem is doing it in a way which isn't a breaking change.
3. Make it so our `SessionsAuthBackend` can be overridden by the user - something like `admin = create_admin(); admin.session_auth_backend.allow_unauthenticated = True`.

What do you think?

I wonder whether using `allow_unauthenticated=True` still requires the user to create the `SessionsBase` and `BaseUser` tables in the database?

**Reply by dantownsend** [reply]
@Skelmis Interested to hear your thoughts on this too.

**Reply by Skelmis** [reply]
I don't think I needed to change too much to modify the admin panel during my initial testing for trying to dockerize the admin panel so I feel like it's plausible without breaking changes

**Reply by sinisaos** [reply]
> 1. Just add allow_authenticated as an argument - but there are already loads of arguments.

@dantownsend This would be the easiest way. Something like this will work

```python
class AdminRouter(FastAPI):

    def __init__(
        ...
        allow_unauthenticated: bool = False,
    ) -> None:
        ...
        self.allow_unauthenticated = allow_unauthenticated
        ...
        backend = (
            SessionsAuthBackend(
                allow_unauthenticated=True,
            )
            if self.allow_unauthenticated is True
            else SessionsAuthBackend(
                auth_table=auth_table,
                session_table=session_table,
                admin_only=True,
                increase_expiry=increase_expiry,
            )
        )

        auth_middleware = partial(
            AuthenticationMiddleware,
            backend=backend,
            on_error=handle_auth_exception,
        )

def create_admin(
    ...
    allow_unauthenticated: bool = False,
):
    
    return AdminRouter(
        ...
        allow_unauthenticated=allow_unauthenticated,
    )

```
Usage would be

```python
create_admin(
    tables=[...],
    allow_unauthenticated=True,
)
```
> I wonder whether using allow_unauthenticated=True still requires the user to create the SessionsBase and BaseUser tables in the database?

No, everything works without providing those tables if `allow_unauthenticated=True`, because in that case we don't use those tables.

**Reply by chilico** [reply]
Thank you for your responses. I had a go at implementing your suggestions by subclassing the `AdminRouter` to accept `allow_unauthenticated` (together with all the other params), then writing a custom `create_admin` to use it. Whilst I was able to get a stable admin panel, I'm still hit with a login form.

**Reply by Skelmis** [reply]
It's fairly okay to patch in, I can't remember exactly if my start on #129 entirely removes the admin panel but the stuff in [this](https://github.com/Skelmis/piccolo_admin/commit/d4115cf3694350709723e42455c87e18ba1818ad) commit can be used to make the admin panel function without authentication, albiet a tad scuffed.

Once I'm back from holiday I'll take the learnings from this to make it use the create admin parameters but yea, that's about what I found was needed

**Reply by sinisaos** [reply]
@chilico I don't know for subclassing `AdminRouter`, but you have two options until it is decided how to allow the user to disable auth.
1. The easiest way is to patch the local Piccolo Admin installation with the code provided in the [previous comment](https://github.com/piccolo-orm/piccolo_admin/discussions/412#discussioncomment-10716307).
2. Another way is to add a custom endpoint file (with the code provided in the [previous comment](https://github.com/piccolo-orm/piccolo_admin/discussions/412#discussioncomment-10716307)) to your project (you also need to add a `dist` file for the admin frontend). After that you can use `create_admin` from that custom endpoint file. In my case both ways work.

[admin.webm](https://github.com/user-attachments/assets/a4fed595-a4cb-4f0d-9308-2367a2516b21)


[Here is an example](https://github.com/sinisaos/simple-piccolo/tree/admin_without_auth) of how to add custom endpoints to an existing project to access Piccolo Admin without auth.
Hope that helps.

**Reply by chilico** [reply]
Ah, thank you @sinisaos. Your second option is a good enough work around for our current needs. But we'll look forward to seeing unauthed access in an upcoming release.
